sdk/java/pulumi/src/main/java/com/pulumi/serialization/internal/ResourcePackages.java

```diff
@@ -1,152 +1,199 @@
 package com.pulumi.serialization.internal;
 
-import com.google.common.base.Strings;
+import com.google.common.base.MoreObjects;
 import com.google.common.base.Suppliers;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
 import com.google.common.reflect.ClassPath;
+import com.google.common.reflect.ClassPath.ClassInfo;
+import com.pulumi.Log;
 import com.pulumi.core.Output;
-import com.pulumi.core.Tuples;
-import com.pulumi.core.Tuples.Tuple2;
+import com.pulumi.core.Urn;
 import com.pulumi.core.annotations.ResourceType;
-import com.pulumi.core.internal.*;
+import com.pulumi.core.internal.Maps;
+import com.pulumi.core.internal.Optionals;
+import com.pulumi.core.internal.PulumiCollectors;
+import com.pulumi.core.internal.Reflection;
+import com.pulumi.core.internal.SemanticVersion;
 import com.pulumi.core.internal.annotations.InternalUse;
-import com.pulumi.resources.*;
+import com.pulumi.resources.ComponentResource;
+import com.pulumi.resources.ComponentResourceOptions;
+import com.pulumi.resources.CustomResource;
+import com.pulumi.resources.CustomResourceOptions;
+import com.pulumi.resources.Resource;
+import com.pulumi.resources.ResourceOptions;
 import pulumirpc.EngineGrpc;
 
 import javax.annotation.Nullable;
 import java.io.IOException;
-import java.lang.reflect.Constructor;
 import java.lang.reflect.InvocationTargetException;
 import java.util.Arrays;
-import java.util.Objects;
 import java.util.Optional;
 import java.util.function.Supplier;
 import java.util.stream.Collectors;
 
+import static com.google.common.base.Strings.isNullOrEmpty;
 import static java.util.Comparator.comparing;
+import static java.util.Objects.requireNonNull;
 import static java.util.stream.Collectors.collectingAndThen;
+import static java.util.stream.Collectors.groupingBy;
 
+@InternalUse
 public class ResourcePackages {
 
-    private static final Supplier<ImmutableMap<String, ImmutableList<Tuple2<Optional<String>, Class<Resource>>>>> resourceTypes =
+    /**
+     * Holds a cache of all Pulumi resources found on the classpath.
+     * The resource type name is the key.
+     */
+    private static final Supplier<ImmutableMap<String, ImmutableList<ResourceTypeEntry>>> resourceTypes =
             Suppliers.memoize(ResourcePackages::discoverResourceTypes); // lazy init
 
-    @SuppressWarnings("UnstableApiUsage")
-    private static ImmutableMap<String, ImmutableList<Tuple2<Optional<String>, Class<Resource>>>> discoverResourceTypes() {
-        final ClassLoader loader = Thread.currentThread().getContextClassLoader();
+    private final Log log;
 
+    @SuppressWarnings("UnstableApiUsage")
+    private static ImmutableMap<String, ImmutableList<ResourceTypeEntry>> discoverResourceTypes() {
+        var loader = MoreObjects.firstNonNull(
+                ResourcePackages.class.getClassLoader(),
```

Comment:
this classloader being first is important for maven, otherwise we get a wrong class loader